### ISSUE TITLE
torch.nn.CTCLoss instead of warp-ctc

### ISSUE THREAD

**Issue by snakers4** [SeanNaren/deepspeech.pytorch#422]
Hi! Has anyone tried using this one as a drop-in replacement in this repo?

In 1.0 it was buggy, but judging by issues / tickets it should have been fixed in 1.1.
I trained my toy models with `torch.nn` loss and it was ok. Will be comparing my benches of warp-ctc against the PyTorch loss shortly.

If anyone has tried this, could you please share your experience.

Some pitfalls
- You need logsoftmax inside of your NN for `torch.nn` loss;
- By default, the PyTorch loss averages;

**Comment by SeanNaren** [comment]
There is still some issues with the built in CTC loss function, this is one for example: https://github.com/pytorch/pytorch/issues/18215

However there is a branch called `ctc_loss` that has implemented the loss function. I'm trying to get it working currently!

**Comment by snakers4** [comment]
So, first of all a couple of things I noticed when porting the changes from the branch / docs:
- Loss has a CUDNN [version](https://pytorch.org/docs/stable/nn.html?highlight=ctc#torch.nn.CTCLoss) (see a bit after loss definition). I could not really make it work even on toy examples some time ago. Maybe it changed;

- @SeanNaren The usage of logsoftmax does not seem to be [correct](https://github.com/SeanNaren/deepspeech.pytorch/blob/ctc_loss/model.py#L71) here - maybe I missed something, but torch.nn loss requires logsoftmax when training, unlike warp-ctc;

- At first when the network exploded, I added some hacks to avoid doing back-prop when loss explodes. At first it seemed that it helped;
```
        if args.max_norm>0:
            if loss_value!=1000 and args.torch_ctc:
                torch.nn.utils.clip_grad_norm_(model.parameters(), args.max_norm)
            if not args.torch_ctc:
                torch.nn.utils.clip_grad_norm_(model.parameters(), args.max_norm)
            
        if torch.isnan(logits).any():
            # work around bad data
            print("WARNING: Skipping NaNs in backward step")
        else:
            # SGD step
            # hack for CTC loss
            if args.torch_ctc:
                if loss_value!=1000 and args.torch_ctc:
                    optimizer.step()
                    if args.enorm: 
                        enorm.step()
                else:
                    print('Not doing optimizer step in case of inf loss')
            else:
                optimizer.step()
                if args.enorm: 
                    enorm.step()  
```

- But then loss continued to explode without any reason
```
GPU-0 Epoch 1 [163/6381]        Time 0.71 (0.74)        Data 0.00 (0.02)        Loss 98.78 (148.39)
WARNING: Working around NaNs in data
GPU-0 Epoch 1 [164/6381]        Time 0.63 (0.74)        Data 0.00 (0.02)        Loss -84.51 (146.97)
```

- I turned off my torch_ctc flag, and everything went back to normal - even no inf loss at all times;

I will try to fiddle with max-norm, maybe it needs to be scaled for torch.nn loss.

**Comment by snakers4** [comment]
Tried different lr / max-norm
Behavior persists
What is interesting - the model diverges when it reaches around ~200th batch

**Comment by YonseiDongEun** [comment]
> Hi! Has anyone tried using this one as a drop-in replacement in this repo?
> 
> In 1.0 it was buggy, but judging by issues / tickets it should have been fixed in 1.1.
> I trained my toy models with `torch.nn` loss and it was ok. Will be comparing my benches of warp-ctc against the PyTorch loss shortly.
> 
> If anyone has tried this, could you please share your experience.
> 
> Some pitfalls
> 
> * You need logsoftmax inside of your NN for `torch.nn` loss;
> * By default, the PyTorch loss averages;

can I ask you how you changed the original ctc loss to torch.nn.CTCLoss() in codes?? 
I try to use it but it makes some errors

**Comment by SeanNaren** [comment]
Just an update, torch.nn.CTCLoss still has a bunch of issues that need to be addressed and can be seen on the PyTorch issue tracker. Once they fix them, we can consider swapping over!

**Comment by Zer0-dev115** [comment]
Can I use CTC loss function in OpenNMT-py?
